### PR TITLE
[skip ci] [WIP] rolling_upgrade: follow MDS upgrade procedure

### PR DESCRIPTION
**Pull request by batrick**
From: http://docs.ceph.com/docs/master/cephfs/upgrading/#upgrading-the-mds-cluster

Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

### CODE REVIEW

**Comment by batrick** [comment]
I don't expect this code to work as-is. I need help rapidly learning how to test this. Also, the procedure is different for pre-Mimic clusters so the code needs to add a check for that and do a slightly differnet procedure:

http://docs.ceph.com/docs/luminous/cephfs/upgrading/

File: infrastructure-playbooks/rolling_update.yml [hunk 1]
```diff
@@ -401,46 +401,35 @@
...
       when: containerized_deployment
 
-    - name: wait for stable fsmap
-      command: ceph --cluster {{ cluster }} fs dump --format=json
-      delay: 5
-      delegate_to: "{{ groups[mon_group_name][0] }}"
-      register: get_fsmap
-      until: {{ get_fsmap.stdout | from_json }}['mds_map']['in'] == [0]
-      when: not containerized_deployment
+    - name: deactivate all non-zero ranks
```
Comment (batrick) [inline infrastructure-playbooks/rolling_update.yml]: @leseb, I need some ansible expertise for this PR. This loop needs to try to repeatedly deactivate the highest rank and then wait for the FSMap to stabilize (i.e. the highest non-zero rank fully deactivates). This requires running a sequence of tasks (mds_pre_upgrade.yml) until a condition is met: `highest_rank == 0`. Unfortunately, it doesn't seem possible to run a sequence of tasks in a loop: there appears to be a bug where `include_tasks` works as expected but the `when:` clause is ignored so the "loop" always runs once.

How should I go about fixing this?

Comment (guits) [reply infrastructure-playbooks/rolling_update.yml]: @batrick according to this issue https://github.com/ansible/ansible/issues/17098 it seems this is still not supported...
At the moment, I don't see an easier workaround than what is proposed by @nerdvegas in the issue

File: infrastructure-playbooks/fs_status.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+- name: get fsmap
+  command: ceph --cluster {{ cluster }} fs dump --format=json
+  delegate_to: "{{ groups[mon_group_name][0] }}"
+  register: get_fsmap
+  when: not containerized_deployment
+
+- name: get fsmap (containerized)
```
Comment (guits) [inline infrastructure-playbooks/fs_status.yml]: Isn't it possible to merge this task with the one above using `{{ docker_exec_cmd }}` logic like we do at many places in the playbook?

File: infrastructure-playbooks/rolling_update.yml (hunk 1 above)
Comment (batrick) [reply infrastructure-playbooks/rolling_update.yml]: Do we have a suggessted approach for rurnning a bash script on the mons? e.g. where do we store the script? Some existing code to look at?

@guits @leseb

File: infrastructure-playbooks/fs_status.yml (hunk 2 above)
Comment (batrick) [reply infrastructure-playbooks/fs_status.yml]: Maybe? I have no idea. I'm just following code I copy-pasted from somewhere else.

File: infrastructure-playbooks/rolling_update.yml (hunk 1 above)
Comment (leseb) [reply infrastructure-playbooks/rolling_update.yml]: You can write a jinja template similar to this https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-defaults/templates/restart_mon_daemon.sh.j2 and then run it like this: https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-defaults/handlers/main.yml#L28

Does that help?

**Comment by leseb** [comment]
@batrick what's the status of this? How can we help? Thanks

**Comment by batrick** [comment]
We need to generate the bash script which can do the loop deactivating MDS ranks. I haven't had time to work on it yet.

**Comment by batrick** [comment]
@leseb @guits I have the CephFS parts of this done but I'm not sure how to get the ansible bits right. Can I get some help? I left some comments in.

There's now a python script (py2/3 compat) that does the preupgrade sequence.

File: infrastructure-playbooks/rolling_update.yml [hunk 3]
```diff
@@ -457,6 +460,29 @@
   become: True
 
   pre_tasks:
+    - name: get fsmap
```
Comment (leseb) [inline infrastructure-playbooks/rolling_update.yml]: move this after the osd upgrade is done or actually create its own section

File: roles/ceph-defaults/handlers/main.yml [hunk 4]
```diff
@@ -24,6 +24,17 @@
   when:
     - mon_group_name in group_names
 
+- name: copy cephfs preupgrade script
```
Comment (leseb) [inline roles/ceph-defaults/handlers/main.yml]: move this into rolling_update.yml

**Comment by leseb** [comment]
@batrick can you allow write access to that PR so I get push into your branch? Thanks! https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

**Comment by batrick** [comment]
@leseb, the PR shows you have permission (that check box is already clicked).

**Comment by leseb** [comment]
@batrick thanks I'll give it a try then.

**Comment by batrick** [comment]
@leseb, if that doesn't work, feel free to just create a new PR.

**Comment by leseb** [comment]
@batrick nop, this hasn't moved yet because I didn't have time yet. Probably next week.

**Comment by batrick** [comment]
ACK, no problem

**Pull request closed without merging**